### PR TITLE
[headless-cache-tuning] Patch 7: --bare + CLAUDE.md injection 

### PR DESCRIPTION
**Pull request by subsetpark**
## Patch 7: --bare + CLAUDE.md injection (gated by ONTON_BARE_CLAUDE)

- Add `?claude_md:string` parameter to lib/prompt.ml render_patch_prompt and a `claude_md_section` template variable. The variable expands to `\n## Project Conventions (CLAUDE.md)\n\n<content>\n` when claude_md is `Some s` and `String.strip s` is non-empty, else empty.
- Slot `{{claude_md_section}}` into the static prefix of the template, immediately after the `# [{{project_name}}]` banner and before `## Problem Statement`. This keeps it inside the cacheable prefix.
- In lib/claude_runner.ml: read `Sys.getenv_opt "ONTON_BARE_CLAUDE"`. When the value is `"1"`, append `"--bare"` to the flag list in build_args and build_stream_args.
- In bin/main.ml at the spawn-site: when `ONTON_BARE_CLAUDE=1`, read the worktree's CLAUDE.md via `Stdlib.In_channel.read_all` (or onton's existing file-read helper) and pass `~claude_md:(Some content)` to render_patch_prompt; otherwise pass `~claude_md:None`.
- Add an inline test in prompt.ml: `claude_md content appears in the static prefix when ~claude_md is Some`.
- Add an inline test: `claude_md section is omitted when ~claude_md is None`.
- Add an inline test in claude_runner.ml: `build_stream_args includes --bare when ONTON_BARE_CLAUDE=1`.
- Add an inline test: `build_stream_args omits --bare when ONTON_BARE_CLAUDE is unset`.

## Changes
- Add `?claude_md:string` parameter to lib/prompt.ml render_patch_prompt and a `claude_md_section` template variable. The variable expands to `\n## Project Conventions (CLAUDE.md)\n\n<content>\n` when claude_md is `Some s` and `String.strip s` is non-empty, else empty.
- Slot `{{claude_md_section}}` into the static prefix of the template, immediately after the `# [{{project_name}}]` banner and before `## Problem Statement`. This keeps it inside the cacheable prefix.
- In lib/claude_runner.ml: read `Sys.getenv_opt "ONTON_BARE_CLAUDE"`. When the value is `"1"`, append `"--bare"` to the flag list in build_args and build_stream_args.
- In bin/main.ml at the spawn-site: when `ONTON_BARE_CLAUDE=1`, read the worktree's CLAUDE.md via `Stdlib.In_channel.read_all` (or onton's existing file-read helper) and pass `~claude_md:(Some content)` to render_patch_prompt; otherwise pass `~claude_md:None`.
- Add an inline test in prompt.ml: `claude_md content appears in the static prefix when ~claude_md is Some`.
- Add an inline test: `claude_md section is omitted when ~claude_md is None`.
- Add an inline test in claude_runner.ml: `build_stream_args includes --bare when ONTON_BARE_CLAUDE=1`.
- Add an inline test: `build_stream_args omits --bare when ONTON_BARE_CLAUDE is unset`.

## Files to Modify
- lib/prompt.ml (modify): Extend render_patch_prompt to accept an optional `?claude_md:string`. When Some, prepend a `## Project Conventions (CLAUDE.md)\n\n<content>` section to the static prefix (just after the project banner, before problem statement) so it lives in the cacheable prefix. When None, omit the section.
- lib/prompt.mli (modify): Add `?claude_md:string` to render_patch_prompt's signature.
- lib/claude_runner.ml (modify): When ONTON_BARE_CLAUDE=1, append --bare to build_args and build_stream_args.
- bin/main.ml (modify): At the spawn site, when ONTON_BARE_CLAUDE=1, read the worktree's CLAUDE.md (if present) and pass to render_patch_prompt as ~claude_md.

## Gameplan Specification

```
module HEADLESS_CACHE_TUNING.

> ══════════════════════════════════════════
> THE GUARANTEE
> ══════════════════════════════════════════
> After this gameplan, every onton-spawned coding-agent
> invocation is tuned for prompt-cache reuse and bounded
> token consumption. Static prompt prefixes are reused
> across patches in one gameplan. Isolated per-spawn
> config dirs stabilise spawn environment. Onton-minted
> session UUIDs remove the parsing race and enable resume
> without re-sending the prompt. Turn caps scale with
> complexity. Per-spawn USD budget caps put a hard floor
> under runaway token spend.

Patch.
Gameplan.
Spawn.
Arg.
Prompt.
Text.
SessionId.
Dir.
Backend.

claude => Backend.
codex => Backend.
pi => Backend.

backend s: Spawn => Backend.
spawn-of p: Patch => Spawn.
complexity p: Patch => Nat.

render p: Patch, gp: Gameplan => Prompt.
prefix pr: Prompt => Text.
suffix pr: Prompt => Text.
shared? t: Text, gp: Gameplan => Bool.
specific? t: Text, p: Patch => Bool.

args s: Spawn => Arg.
has-flag? a: Arg, name: String => Bool.

config-dir s: Spawn => Dir.
max-turns s: Spawn => Nat.

session-id p: Patch => SessionId.
persisted-before-spawn? id: SessionId, s: Spawn => Bool.
passed-to-cli? id: SessionId, s: Spawn => Bool.
minted-session-flag-on? => Bool.

budget-cap-set? s: Spawn => Bool.
flag-emitted? s: Spawn => Bool.
cost-tracked? s: Spawn => Bool.
self-terminates-on-overshoot? s: Spawn => Bool.

bare-flag-on? => Bool.
worktree-claude-md p: Patch => Text.
non-empty? t: Text => Bool.
prompt-of s: Spawn => Prompt.
contains? t: Text, sub: String => Bool.
---

> Static prefix is gameplan-shared.
all p: Patch, gp: Gameplan | shared? (prefix (render p gp)) gp.

> Dynamic suffix is patch-specific.
all p: Patch, gp: Gameplan | specific? (suffix (render p gp)) p.

> Claude spawns pass --exclude-dynamic-system-prompt-sections.
all s: Spawn, backend s = claude |
  has-flag? (args s) "--exclude-dynamic-system-prompt-sections".

> Distinct patches have distinct config dirs.
all p1: Patch, p2: Patch, p1 ~= p2 |
  config-dir (spawn-of p1) ~= config-dir (spawn-of p2).

> Turn caps are monotone non-decreasing in complexity.
all p1: Patch, p2: Patch, complexity p1 < complexity p2 |
  max-turns (spawn-of p1) <= max-turns (spawn-of p2).

> Session-ID minting (when enabled) persists before spawn and
> passes to the CLI.
all p: Patch, minted-session-flag-on? |
  persisted-before-spawn? (session-id p) (spawn-of p).

all p: Patch, minted-session-flag-on? |
  passed-to-cli? (session-id p) (spawn-of p).

> When the budget cap is set, Claude spawns emit the flag and
> Codex spawns track cost + self-terminate.
all s: Spawn, backend s = claude, budget-cap-set? s |
  flag-emitted? s.

all s: Spawn, backend s = codex, budget-cap-set? s |
  cost-tracked? s and self-terminates-on-overshoot? s.

> When --bare is enabled, Claude spawns pass --bare and the
> worktree's CLAUDE.md content is reachable from the spawn's
> prompt prefix.
all p: Patch, bare-flag-on?, backend (spawn-of p) = claude |
  has-flag? (args (spawn-of p)) "--bare".

all p: Patch, bare-flag-on?, non-empty? (worktree-claude-md p) |
  contains? (prefix (prompt-of (spawn-of p))) "CLAUDE.md".

```

## Patch Specification

```
module HEADLESS_CACHE_TUNING_PATCH_7.

> Patch 8 hermeticises Claude spawns: --bare drops auto-loaded
> hooks/plugins/MCP/CLAUDE.md from the agent's ambient state,
> and onton explicitly prepends the worktree's CLAUDE.md as a
> static section so project conventions still reach the agent.
> Net effect: identical agent context across machines.

Patch.
Spawn.
Backend.
Prompt.
Text.

claude => Backend.
backend s: Spawn => Backend.
spawn-of p: Patch => Spawn.
prompt-of s: Spawn => Prompt.
prefix pr: Prompt => Text.
contains? t: Text, sub: String => Bool.
has-flag? s: Spawn, name: String => Bool.
bare-flag-on? => Bool.
worktree-claude-md p: Patch => Text.
non-empty? t: Text => Bool.
---

> When the bare flag is on, Claude spawns pass --bare.
all p: Patch, bare-flag-on?, backend (spawn-of p) = claude |
  has-flag? (spawn-of p) "--bare".

> When the bare flag is on and the worktree has a CLAUDE.md,
> its content reaches the spawn's prompt prefix.
all p: Patch, bare-flag-on?, non-empty? (worktree-claude-md p) |
  contains? (prefix (prompt-of (spawn-of p))) "CLAUDE.md".

```

## Implementation Notes

- Deviates from the original rollout plan in two ways: the `ONTON_BARE_CLAUDE` gate was removed after implementation review, and the injected conventions file is `AGENTS.md` rather than `CLAUDE.md`. Claude now always runs with `--bare`, and the spawn path injects worktree `AGENTS.md` when present.
- `render_patch_prompt` still keeps `?agents_md` optional, which avoids churn at unrelated call sites and keeps the change localized to the Claude spawn path.
- The spawn-site file read uses the repo's existing `open_text` + `input_all` pattern rather than `Stdlib.In_channel.read_all`, because that function is not available in this OCaml stdlib.

## Summary by CodeRabbit

* **New Features**
  * Added `ONTON_BARE_CLAUDE` environment variable support to enable Claude's bare flag mode when processing patches
  * Added support for optional `CLAUDE.md` configuration file that can be placed in the worktree directory to customize Claude behavior and settings

* **Chores**
  * Improved terminal signal handling and initialization

### CODE REVIEW

File: lib/claude_runner.ml [hunk 1]
```diff
@@ -53,13 +53,17 @@ let model_args = function
   | Some m when not (String.is_empty m) -> [ "--model"; m ]
   | _ -> []
 
+let bare_args = [ "--bare" ]
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/claude_runner.ml:77]: Restored the ONTON_BARE_CLAUDE gate. Claude now appends `--bare` only when that env var is exactly `1`, and the spawn site only injects worktree `AGENTS.md` into the prompt when the same gate is enabled.

File: lib/prompt.ml [hunk 2]
```diff
@@ -95,8 +95,13 @@ let optional_list_section ~header items =
   | [] -> ""
   | _ -> optional_section ~header (format_list items)
 
-let render_patch_prompt ~(project_name : string) ?pr_number (patch : Patch.t)
-    (gameplan : Gameplan.t) ~(base_branch : string) =
+let agents_md_section = function
+  | Some content when not (String.is_empty (String.strip content)) ->
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:99]: Reverted this patch back to `CLAUDE.md` so the implementation matches the current gameplan/patch spec exactly. The spawn site now reads `CLAUDE.md`, the injected header is `Project Conventions (CLAUDE.md)`, and the inline prompt test asserts that `CLAUDE.md` is present in the static prefix.

File: lib/claude_runner.ml [hunk 3]
```diff
@@ -56,6 +56,9 @@ let model_args = function
 let max_turns_for ~complexity =
   match complexity with Some 1 -> 50 | Some 2 -> 100 | _ -> 200
 
+let bare_args ~getenv_opt =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/claude_runner.ml]: Threaded `~getenv_opt` through both `build_args` and `build_stream_args`, with `Sys.getenv_opt` still as the default. The new inline tests now call `build_stream_args` itself with injected env lookups, so they exercise the real integration path instead of only testing `bare_args` in isolation.

File: bin/main.ml [hunk 4]
```diff
@@ -502,6 +502,16 @@ let read_artifact_file path =
     else None
   with _ -> None
 
+let read_optional_file path =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml:489]: Consolidated the two file-read helpers so `read_artifact_file` now aliases the same `In_channel.open_text`/`input_all` implementation used by `read_optional_file`. That removes the API inconsistency and the duplicated read logic.

File: lib/prompt.ml [hunk 5]
```diff
@@ -241,6 +247,7 @@ Continue implementing until all tests pass.|}
   render_with_override ~project_name ~name:"patch" ~vars ~default:(fun () ->
       substitute_variables
         {|# [{{project_name}}]
+{{agents_md_section}}
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Normalized the template spacing by removing the section leading newline and making the template own the single blank line after the banner. The CLAUDE.md section now contributes only its own body plus a trailing blank line, so the rendered spacing is consistent whether the optional section is present or absent.

File: lib/prompt.ml [hunk 6]
```diff
@@ -95,8 +95,13 @@ let optional_list_section ~header items =
   | [] -> ""
   | _ -> optional_section ~header (format_list items)
 
-let render_patch_prompt ~(project_name : string) ?pr_number (patch : Patch.t)
-    (gameplan : Gameplan.t) ~(base_branch : string) =
+let claude_md_section = function
+  | Some content when not (String.is_empty (String.strip content)) ->
+      "## Project Conventions (CLAUDE.md)\n\n" ^ content ^ "\n\n"
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Applied this cleanup on current HEAD.  now uses  before appending the final blank line, and I added an inline test covering a trailing-newline input so the rendered spacing stays exactly consistent.

File: bin/main.ml [hunk 7]
```diff
@@ -2046,8 +2044,20 @@ let runner_fiber ~runtime ~env ~config ~pick_backend ~project_name ~pr_registry
...
+                                      let claude_md =
+                                        match
+                                          Stdlib.Sys.getenv_opt
+                                            "ONTON_BARE_CLAUDE"
+                                        with
+                                        | Some "1" ->
+                                            read_optional_file
+                                              (Stdlib.Filename.concat _wt_path
+                                                 "CLAUDE.md")
+                                        | _ -> None
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml]: Addressed in later commits after . Current HEAD reads  at the spawn site and labels the injected section , so the path/header/spec are aligned again.

[the remark this answers is not in the document]
Comment (subsetpark) [reply bin/main.ml]: Addressed in later commits after ececff2. Current HEAD reads CLAUDE.md at the spawn site and labels the injected section Project Conventions (CLAUDE.md), so the path, header, and spec are aligned again.

File: lib/prompt.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Applied this cleanup on current HEAD. The helper now uses String.rstrip on the file contents before appending the final blank line, and there is an inline test covering input that already ends with a newline.

File: lib/claude_runner.ml [hunk 8]
```diff
@@ -563,6 +589,24 @@ let%test "prepare_minted_session_id errors when flag is on and path missing" =
   | Error _ -> true
   | Ok _ -> false
 
+let%test "build_stream_args includes --bare when ONTON_BARE_CLAUDE=1" =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/claude_runner.ml]: Added the matching non-streaming coverage on current HEAD. There are now explicit build_args tests for both ONTON_BARE_CLAUDE=1 and the unset case, so both Claude invocation paths are covered if bare_args is refactored later.

File: lib/term.ml [hunk 9]
```diff
@@ -397,6 +397,11 @@ module Raw = struct
 
   let _saved_winch_handler : Stdlib.Sys.signal_behavior option ref = ref None
 
+  (* [Sys] does not expose [sigwinch] on every toolchain. Linux and Darwin use
+     signal number 28 for terminal resize notifications, which are the Unix
+     targets supported by this TUI module. *)
+  let sigwinch = 28
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/term.ml]: Reverted this on current HEAD. The portability rationale for hardcoding 28 was too weak for the project target matrix, so lib/term.ml now uses Stdlib.Sys.sigwinch again.
